### PR TITLE
Sema: Fixed a crash when accessing a concrete associated type through an existential metatype

### PR DESCRIPTION
**Pull request by sl**
Fixed a bug that caused a crash when accessing a concrete typealias through an existential metatype.

Additionally added tests for access to concrete typealiases both through existential metatype and protocol metatypes.

Resolves [SR-9781](https://bugs.swift.org/browse/SR-9781).

### CODE REVIEW

File: test/type/metatype/metatypes.swift [hunk 1]
```diff
@@ -18,3 +18,13 @@ var test1 = Test1a.self
...
 test1 = type(of: x)
+
+// test associated type access through metatypes
+protocol P {
+  typealias E = Int
+}
+
+func foo(meta1: P.Protocol, meta2: P.Type) {
+  print(meta1.E.self)
+  print(meta2.E.self)
```
Comment (slavapestov) [inline test/type/metatype/metatypes.swift]: Can you test what happens if the typealias underlying type depends on Self or an associated type?

File: lib/Sema/CSApply.cpp [hunk 2]
```diff
@@ -6453,6 +6453,12 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
       }
     }
 
+    // Allows for same-type coercion with existential metatypes.
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp]: It’s a bit odd that coerceToType is called like this. Is toMeta an existential metatype or protocol metatype?

File: lib/AST/ASTVerifier.cpp [hunk 3]
```diff
@@ -1816,15 +1816,6 @@ class Verifier : public ASTWalker {
         abort();
       }
       
-      // The base of a member reference cannot be an existential type.
-      if (E->getBase()->getType()->getWithoutSpecifierType()
```
Comment (slavapestov) [inline lib/AST/ASTVerifier.cpp]: Maybe this should remain but isAnyExistentialType can be tightened up into isExistentialType (It’s still invalid to access a member type of an existential value)

Comment (sl) [reply lib/AST/ASTVerifier.cpp]: I noticed the same thing, was just adding that now, will have a commit up in a second.

File: lib/Sema/CSApply.cpp (hunk 2 above)
Comment (sl) [reply lib/Sema/CSApply.cpp]: Ah, I think you're right that toMeta is a protocol metatype here, so I might need to change the way I handle this.

**Comment by sl** [comment]
@slavapestov Thanks for pointing out some issues with this patch. I'm going to take a look at this tomorrow and try to figure out the correct way to address this.

**Comment by sl** [comment]
Reopening this with updated change.

**Comment by sl** [comment]
Reopening this with updated change.

File: test/Constraints/protocols.swift [hunk 4]
```diff
@@ -216,6 +217,12 @@ func staticExistential(_ p: P.Type, pp: P.Protocol) {
   // Static member of metatype -- not allowed
   _ = pp.tum // expected-error{{static member 'tum' cannot be used on protocol metatype 'P.Protocol'}}
   _ = P.tum // expected-error{{static member 'tum' cannot be used on protocol metatype 'P.Protocol'}}
+
+  // Concrete associated type access through protocol metatype
+  _ = pp.E.self
```
Comment (AnthonyLatsis) [inline test/Constraints/protocols.swift]: Nit: In this case `E` is simply a `typealias`, a concrete associated type is an actual associated type that has been same-type constrained.

Comment (sl) [reply test/Constraints/protocols.swift]: I think I might have been a bit too tired when I wrote this comment, nice spot.

**Comment by slavapestov** [comment]
It looks like the change might be incomplete - did you accidentally force push the wrong version by any chance?

**Comment by AnthonyLatsis** [comment]
@slavapestov This is my fault. I have lead Sam astray on JIRA by pointing him to member lookup; I was under the impression this is [SR-9784](https://bugs.swift.org/browse/SR-9784). I'm sorry @sl, I should have taken a better look at what you originally had before advising anything.. We want to *allow* accessing typealiases on existential metatypes, so it happens your initial approach was closer to the real solution.

**Comment by AnthonyLatsis** [comment]
> It’s a bit odd that coerceToType is called like this. Is toMeta an existential metatype or protocol metatype?

The coercion is failing from an existential metatype to simply a metatype. Maybe we should instead be opening the existential metatype in [`getTypeOfMemberReference`](https://github.com/apple/swift/blob/e922d2923bc45b6739a698cab8080d0b9a014e6a/lib/Sema/ConstraintSystem.cpp#L1211)?

**Comment by sl** [comment]
Ah, that makes sense. No worries about the confusion. Once I get home, I'll revert back to the old version of the pull request and continue work from there.

Thanks to both of you for following along with this PR 🙂

**Comment by slavapestov** [comment]
> The coercion is failing from an existential metatype to simply a metatype.

Yeah, that's not a valid coercion, so something has gone wrong further upstream before you even get to coerceToType().

> Maybe we should instead be opening the existential metatype in getTypeOfMemberReference?

Maybe. In practice the concrete type inside the existential metatype doesn't matter since the member is just a type alias that doesn't reference any associated types... but if its simpler overall to be consistent we should make sure we open it when accessing a member type.

**Comment by sl** [comment]
@slavapestov @AnthonyLatsis Just pushed a version that fixes the issue by preventing the invalid coercion in the case of an existential metatype. If you could take a look at it when you get the chance, that'd be great.

Thanks again for following this PR so closely.

**Review by slavapestov — APPROVED** [review]
Looks good!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by sl** [comment]
Looks like something’s going wrong in the subclass_existentials tests, I’ll resolve any test failures, then smoke again in a bit.

**Comment by sl** [comment]
@swift-ci Please smoke test

**Comment by sl** [comment]
@slavapestov Had to add an additional check to make sure the coercion was actually invalid. Just wanted to make sure this is still good to merge.

**Comment by AnthonyLatsis** [comment]
Thanks, Sam!

**Review by slavapestov — APPROVED** [review]

**Pull request merged**